### ISSUE TITLE
(Literal[True]) -> None" cannot be assigned to return type "(bool) -> None"

### ISSUE THREAD

**Issue by dbrattli** [microsoft/pyright#2796]
**Describe the bug**

Came across this inconsistency that seems like a bug. The return statement of `arrow_16` gives error `(Literal[True]) -> None" cannot be assigned to return type "(bool) -> None"`. But assigning the return value to a variable `ret` first fixes the error. This seems inconsistent (whatever the error).

**To Reproduce**
Steps to reproduce the behavior.

```py
from typing import Callable, TypeVar

_T = TypeVar("_T")

def Return(value: _T) -> Callable[[_T], None]:
    def ret(ctx: _T):
        pass

    return ret

def arrow_168() -> Callable[[bool], None]:
    return Return(True)
```

<img width="915" alt="Screenshot 2022-01-03 at 16 13 09" src="https://user-images.githubusercontent.com/849479/147947111-73790961-8a66-4358-b0fe-3c920ec9abd7.png">

**Expected behavior**
A clear and concise description of what you expected to happen. However, assigning the return value to a variable `ret` before returning removes the error:

```py
from typing import Callable, TypeVar

_T = TypeVar("_T")

def Return(value: _T) -> Callable[[_T], None]:
    def ret(ctx: _T):
        pass

    return ret

def arrow_168() -> Callable[[bool], None]:
    ret = Return(True)
    return ret
```

I would expect both examples to produce the same type check result with pyright.

**VS Code extension or command-line**

Pyright: v1.1.203, vscode insider

**Comment by erictraut** [comment]
Pyright contains some complex heuristics in the TypeVar constraint solver for determining whether to retain literals when solving the constraints. In this case, these heuristics were incorrectly determining that it should retain literals, so it solved the constraints as `_T = Literal[True]` rather than `_T = bool`. This occurred only in the case where bidirectional type inference was used, hence why assigning it to a temporary variable made the error go away.

I've updated the heuristics so this code works. The change will be in the next release.

**Comment by erictraut** [comment]
This is included in pyright 1.1.204, which I just published. It will also be included in the next release of pylance.

**Comment by dbrattli** [comment]
Thanks for fixing. There's however still some problems with this as can be seen for `singleton.Return` below, so I have updated my minimal repro to be a little more complex.

<img width="1372" alt="Screenshot 2022-01-09 at 09 16 12" src="https://user-images.githubusercontent.com/849479/148674689-d10ab62d-abcb-42d0-8c67-1c1131006f46.png">

```py
from typing import Callable, TypeVar, Generic

_T = TypeVar("_T")

class IAsyncContext(Generic[_T]):
    pass

Async = Callable[[IAsyncContext[_T]], None]

def Return(value: _T) -> Async[_T]:
    def ret(ctx: IAsyncContext[_T]) -> None:
        pass

    return ret

def arrow_3() -> Async[bool]:
    return Return(True)
```

This now gives the same error as before.

**Comment by erictraut** [comment]
Thanks for the updated sample. Yeah, this variant eluded my heuristic update because it involves a function parameter, therefore it's contravariant. I've further updated the heuristics to handle this case better.

**Comment by dbrattli** [comment]
I have one more that is failing if the returned value is a tuple:

```py
from typing import Callable, TypeVar, Generic, Tuple

_T = TypeVar("_T")

class IAsyncContext(Generic[_T]):
    pass

Async = Callable[[IAsyncContext[_T]], None]

def Return(value: _T) -> Async[_T]:
    def ret(ctx: IAsyncContext[_T]) -> None:
        pass

    return ret

def arrow_3() -> Async[Tuple[None, bool]]:
    return Return((None, True))
```

**Comment by erictraut** [comment]
Thanks for reporting this, but could you please open a new issue? This one is a somewhat different problem.
